### DISCUSSION TITLE
Time Varying System Matrices

### DISCUSSION THREAD

**Discussion by trmthcode3** [do-mpc/do-mpc#322] [Q&A] [answered]
So I am looking at the following example: https://www.do-mpc.com/en/latest/example_gallery/oscillating_masses_discrete.html.

In this example, the system is defined by its A and B matrices. I have something similar, however, the A and B matrices are not constant valued.

I am wondering how can I define a system by system matrices that are scheduling parameter dependent (LPV system). Would I have to setup each entry of the matrix as a tvp (time varying parameter)? Or would it be better/possible to load in the entire matrix depending on the state of the system? This sounds like it could be difficult and I am not sure how I could build a model based on some thing that parses matrices at each new state. It seems more likely to setup each entry as a tvp but I would love to hear some thoughts/advice about this.

Thanks in advance!

**Comment by 4flixt** [comment, marked as answer]
Hi @trmthcode3 

are your (time-varying) matrices A and B functions of the states, inputs, etc? In that case, you wont need time-varying parameters. If they are functions of some other quantity that you cannot influence but for which a prediction is available (e.g. ambient temperature) then you will need time-varying parameters. 

In both cases you have to consider that do-mpc will require a classical ODE formulation with one differential equation for each state
These equations can be read directly from your system matrices. It should be straight-forward to encode this.

Best,
4flixt

**Reply by trmthcode3** [reply]
Thanks for the reply. The matrices are a function of the scheduling variables which happen to be two of the five state variables.

I should add, the model is a learned model. So I can give a neural network the current state (and corresponding scheduling variables at any given time step) and it will give the two matrices at that time step and then use that to calculate the state at the next given time step.

Can you clarify what exactly you mean by reading the differential equation from the system matrices? The model is discrete so it is in the form of x_k+1 = A*x_k + B*u_k, so I don't really have a classical ODE available?

It seems that all the examples on the do-mpc page don't consider A and B matrices for the ODE formulations. I have only seen the A and B matrices considered for discrete systems?

**Reply by 4flixt** [reply]
Do-MPC can handle discrete equations. You just have to initialize the model with the respective keyword. 
And yes, do-mpc doesn't have an interface to pass A, B, C, D matrices regardless if you have a discrete or continuous model. 
Can you please clarify for yourself what the expression: 
x_k+1 = A x_k + B u_k means? 
I think you are lacking some fundamental understanding here that has nothing to do with do-mpc.

**Reply by trmthcode3** [reply]
The x_k+1 = A x_k + B u_k is similar to what is used on the oscillating masses example, which my understanding is just a discrete realization of a state space model:
<img width="617" alt="Screen Shot 2022-11-09 at 6 12 50 PM" src="https://user-images.githubusercontent.com/63681318/200962003-ab9126f8-76f2-48af-8600-a10a015ca642.png">
The screenshot attached is directly from that example.

I am confused when you say do-mpc doesn't have an interface to pass A, B, C, D matrices because isn't that exactly what the oscillating masses example is doing?
<img width="752" alt="Screen Shot 2022-11-09 at 6 14 15 PM" src="https://user-images.githubusercontent.com/63681318/200962215-d8752f05-c4c5-4ee1-95d3-ac7bc08d35dd.png">

**Reply by trmthcode3** [reply]
> Do-MPC can handle discrete equations. You just have to initialize the model with the respective keyword. And yes, do-mpc doesn't have an interface to pass A, B, C, D matrices regardless if you have a discrete or continuous model. Can you please clarify for yourself what the expression: x_k+1 = A x_k + B u_k means? I think you are lacking some fundamental understanding here that has nothing to do with do-mpc.

Could you still clarify what you meant by "do-mpc doesn't have an interface to pass A, B, C, D matrices regardless if you have a discrete or continuous model"? Because as I showed in that example, I see the matrices being used. It is with numpy yes, but I am still unsure what was meant by the comment.

**Reply by 4flixt** [reply]
do-mpc is based on the concept of introducing states (`x`) and then defining a right-hand side function `f(x,u)`which either represents a differential equation for this state (continuous model) or the state at the next time step. 
do-mpc also allows for these states to be vectors or even matrices. 
So in this example, we introduce a single state (which is a vector representing all states of the sytem) and we set a single right-hand side equation. 
In this example, the right-hand side equation is the classical:

```
x_next = A@x + B@u
```

but it could also be anything else. You are not directly passing the matrices `A` and `B` to do-mpc. It doesn't have an interface for it. You are passing the right-hand-side equation. 
Also notice that only in this example with a single (vector) state we have a single right-hand-side equation which can be written in the form above.

**Reply by trmthcode3** [reply]
I see so it's not that the system can't be defined by A and B matrices, it is just that the package doesn't have a function to just input those matrices and run the model from there. Basically, you can use the matrices to define the right-hand side function, but you can't just give the matrices directly to some function/method in the package?

Also, when you said "Can you please clarify for yourself what the expression:
x_k+1 = A x_k + B u_k means?
I think you are lacking some fundamental understanding here that has nothing to do with do-mpc."

Was this: "The x_k+1 = A x_k + B u_k is similar to what is used on the oscillating masses example, which my understanding is just a discrete realization of a state space model" a reasonable response? I am not sure by what you meant by a lacking in fundamental understanding of this equation as to me I just see it as the discrete time state space model. I think maybe I was getting confused with your very first response when you said a classical ODE is needed, I interpreted that as you saying we needed a differential equation not in the typical Ax+Bu state space form, which is why I said I didn't have one available. Perhaps that's where the supposed lack of understanding was seen? Reading back the initial response makes more sense now.

Finally, after seeing what you mean by the matrices, essentially I can change the right hand side whenever I want correct? So I could redefine the RHS function when a certain point in the state space is reached and thus the scheduling variable change?

**Reply by 4flixt** [reply]
Hi @trmthcode3,

👍  to your first question.

Regarding my second comment: Assume you have something like this:
```
A = array([[2, 9, 4],
       [9, 5, 2],
       [6, 3, 7]])
```
then you can write the right-hand-side (be it a continuous or discrete model) as `A@x` (I ignore `B` for simplicity).
But what that actually means is:
```
x_next[0] = 2*x[0]+9*x[1]+4*x[2]
x_next[1] = 9*x[0]+5*x[1]+2*x[2]
x_next[2] = 6*x[0]+3*x[1]+7*x[2]
```
In do-mpc you will often write down these equations explicitly and this is what I meant by "clarify for yourself what x_k+1 = A x_k + B u_k means".

If you write down the equations in this form you can give states explicit names (representing their physical quantity)and you can have have arbitrary (non-linear) expressions for each state. If your `A` matrix is time-varying this simply means your expressions are non-linear. 

Best,
4flixt
